Crash program when I call getGeners

I get  this error

```
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): Failed to handle method call
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): java.lang.NullPointerException: Attempt to invoke interface method 'void io.flutter.plugin.common.MethodChannel$Result.success(java.lang.Object)' on a null object reference
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at com.ryanheise.audioservice.AudioServicePlugin.sendStartResult(AudioServicePlugin.java:143)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at com.ryanheise.audioservice.AudioServicePlugin.access$800(AudioServicePlugin.java:54)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at com.ryanheise.audioservice.AudioServicePlugin$BackgroundHandler.onMethodCall(AudioServicePlugin.java:617)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:231)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:93)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:642)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at android.os.MessageQueue.next(MessageQueue.java:376)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at android.os.Looper.loop(Looper.java:244)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at android.app.ActivityThread.main(ActivityThread.java:6706)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/MethodChannel#ryanheise.com/audioServiceBackground( 8672): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
E/DartMessenger( 8672): Uncaught exception in binary message listener
E/DartMessenger( 8672): java.lang.IllegalStateException: Reply already submitted
E/DartMessenger( 8672): 	at io.flutter.embedding.engine.dart.DartMessenger$Reply.reply(DartMessenger.java:148)
E/DartMessenger( 8672): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:249)
E/DartMessenger( 8672): 	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(DartMessenger.java:93)
E/DartMessenger( 8672): 	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(FlutterJNI.java:642)
E/DartMessenger( 8672): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/DartMessenger( 8672): 	at android.os.MessageQueue.next(MessageQueue.java:376)
E/DartMessenger( 8672): 	at android.os.Looper.loop(Looper.java:244)
E/DartMessenger( 8672): 	at android.app.ActivityThread.main(ActivityThread.java:6706)
E/DartMessenger( 8672): 	at java.lang.reflect.Method.invoke(Native Method)
E/DartMessenger( 8672): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
E/DartMessenger( 8672): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
Error -32601 received from application: Method not found
Error -32601 received from application: Method not found
Restarted application in 1,769ms.
method not available: ext.flutter.platformOverride
Reloaded 3 of 757 libraries in 857ms.
Reloaded 3 of 757 libraries in 779ms.

[38;5;248m════════ Exception caught by widgets library ═══════════════════════════════════[39;49m
[38;5;244mThe following assertion was thrown building Library(dirty, state: _LibraryState#71a6f):[39;49m
'dart:ui/painting.dart': Failed assertion: line 188: '<optimized out>': is not true.


[38;5;248mEither the assertion indicates an error in the framework itself, or we should provide substantially more information in this error message to help you determine and fix the underlying cause.
In either case, please report this assertion by filing a bug on GitHub:
  https://github.com/flutter/flutter/issues/new?template=BUG.md
[39;49m

[38;5;244mThe relevant error-causing widget was[39;49m
    [38;5;248mLibrary[39;49m
[38;5;244mWhen the exception was thrown, this was the stack[39;49m
[38;5;244m#2      Color.withOpacity  (dart:ui/painting.dart:188:12)[39;49m
[38;5;248m#3      _LibraryState.build[39;49m
[38;5;244m#4      StatefulElement.build[39;49m
[38;5;244m#5      ComponentElement.performRebuild[39;49m
[38;5;244m#6      Element.rebuild[39;49m
[38;5;244m...[39;49m
[38;5;248m════════════════════════════════════════════════════════════════════════════════[39;49m
Reloaded 3 of 757 libraries in 922ms.
Reloaded 3 of 757 libraries in 877ms.
Reloaded 3 of 757 libraries in 816ms.
E/AndroidRuntime( 8672): FATAL EXCEPTION: AsyncTask #40
E/AndroidRuntime( 8672): Process: com.example.musical, PID: 8672
E/AndroidRuntime( 8672): java.lang.RuntimeException: An error occurred while executing doInBackground()
E/AndroidRuntime( 8672): 	at android.os.AsyncTask$3.done(AsyncTask.java:354)
E/AndroidRuntime( 8672): 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
E/AndroidRuntime( 8672): 	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
E/AndroidRuntime( 8672): 	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
E/AndroidRuntime( 8672): 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
E/AndroidRuntime( 8672): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/AndroidRuntime( 8672): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/AndroidRuntime( 8672): 	at java.lang.Thread.run(Thread.java:764)
E/AndroidRuntime( 8672): Caused by: android.database.sqlite.SQLiteException: no such column: genre_name (code 1 SQLITE_ERROR): , while compiling: SELECT Distinct genre_name FROM audio ORDER BY genre_name ASC
E/AndroidRuntime( 8672): 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:179)
E/AndroidRuntime( 8672): 	at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:135)
E/AndroidRuntime( 8672): 	at android.content.ContentProviderProxy.query(ContentProviderNative.java:418)
E/AndroidRuntime( 8672): 	at android.content.ContentResolver.query(ContentResolver.java:804)
E/AndroidRuntime( 8672): 	at android.content.ContentResolver.query(ContentResolver.java:753)
E/AndroidRuntime( 8672): 	at android.content.ContentResolver.query(ContentResolver.java:711)
E/AndroidRuntime( 8672): 	at boaventura.com.devel.br.flutteraudioquery.loaders.GenreLoader$GenreLoadTask.loadData(GenreLoader.java:111)
E/AndroidRuntime( 8672): 	at boaventura.com.devel.br.flutteraudioquery.loaders.GenreLoader$GenreLoadTask.loadData(GenreLoader.java:86)
E/AndroidRuntime( 8672): 	at boaventura.com.devel.br.flutteraudioquery.loaders.tasks.AbstractLoadTask.doInBackground(AbstractLoadTask.java:45)
E/AndroidRuntime( 8672): 	at boaventura.com.devel.br.flutteraudioquery.loaders.tasks.AbstractLoadTask.doInBackground(AbstractLoadTask.java:12)
E/AndroidRuntime( 8672): 	at android.os.AsyncTask$2.call(AsyncTask.java:333)
E/AndroidRuntime( 8672): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/AndroidRuntime( 8672): 	... 4 more
I/Process ( 8672): Sending signal. PID: 8672 SIG: 9
Lost connection to device.
Exited (sigterm)
```

Same error here, in Android 10. I get the following:

```
E/AndroidRuntime(23721): FATAL EXCEPTION: AsyncTask #1
E/AndroidRuntime(23721): Process: com.pedrolemoz.music_player, PID: 23721
E/AndroidRuntime(23721): java.lang.RuntimeException: An error occurred while executing doInBackground()
E/AndroidRuntime(23721):        at android.os.AsyncTask$4.done(AsyncTask.java:399)
E/AndroidRuntime(23721):        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
E/AndroidRuntime(23721):        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
E/AndroidRuntime(23721):        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
E/AndroidRuntime(23721):        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
E/AndroidRuntime(23721):        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/AndroidRuntime(23721):        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/AndroidRuntime(23721):        at java.lang.Thread.run(Thread.java:919)
E/AndroidRuntime(23721): Caused by: android.database.sqlite.SQLiteException: no such column: genre_name (code 1 SQLITE_ERROR): , while compiling: SELECT Distinct genre_name FROM audio WHERE ((is_pending=0) AND (is_trashed=0) AND (volume_name IN ( 'external_primary' ))) ORDER BY genre_name ASC
E/AndroidRuntime(23721):        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:184)
E/AndroidRuntime(23721):        at android.database.DatabaseUtils.readExceptionFromParcel(DatabaseUtils.java:140)
E/AndroidRuntime(23721):        at android.content.ContentProviderProxy.query(ContentProviderNative.java:423)
E/AndroidRuntime(23721):        at android.content.ContentResolver.query(ContentResolver.java:946)
E/AndroidRuntime(23721):        at android.content.ContentResolver.query(ContentResolver.java:881)
E/AndroidRuntime(23721):        at android.content.ContentResolver.query(ContentResolver.java:837)
E/AndroidRuntime(23721):        at boaventura.com.devel.br.flutteraudioquery.loaders.GenreLoader$GenreLoadTask.loadData(GenreLoader.java:111)
E/AndroidRuntime(23721):        at boaventura.com.devel.br.flutteraudioquery.loaders.GenreLoader$GenreLoadTask.loadData(GenreLoader.java:86)
E/AndroidRuntime(23721):        at boaventura.com.devel.br.flutteraudioquery.loaders.tasks.AbstractLoadTask.doInBackground(AbstractLoadTask.java:45)
E/AndroidRuntime(23721):        at boaventura.com.devel.br.flutteraudioquery.loaders.tasks.AbstractLoadTask.doInBackground(AbstractLoadTask.java:12)
E/AndroidRuntime(23721):        at android.os.AsyncTask$3.call(AsyncTask.java:378)
E/AndroidRuntime(23721):        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/AndroidRuntime(23721):        ... 4 more
I/Process (23721): Sending signal. PID: 23721 SIG: 9
Lost connection to device.
```

@sc4v3ng3r Please merge pull requests to this package!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,10 @@
-## 0.3.5+2
+## 0.3.5+5
  * Support loading artwork on Android >= Q with getArtwork method
- 
+ * **Bug fix**: getGenres call bug fixed
+ * **Bug fix**: getSongsById when call with one single id
+ * **Bug fix**: getSongsFromArtist call
+ * **Breaking change**:
+    * The method getSongsFromArtist now accepts artistId as parameter.
 
 ## 0.3.4+2  
 * Support V2 embedding Flutter API.

--- a/README.md
+++ b/README.md
@@ -49,7 +49,7 @@ To use this plugin, add `flutter_audio_query` as a [dependency in your pubspec.y
 
 ```yaml
   dependencies:
-    flutter_audio_query: "^0.3.4+1"
+    flutter_audio_query: "^0.3.5+5"
 ```
 
 ## API

--- a/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/ArtistLoader.java
+++ b/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/ArtistLoader.java
@@ -115,6 +115,7 @@ public class ArtistLoader extends AbstractLoader {
                                ArtistSortType sortType){
         String[] selectionArgs;
         String sortOrder = null;
+        String selection = MediaStore.Audio.Artists._ID;
 
         if (ids == null || ids.isEmpty()) {
             result.error("NO_ARTIST_IDS", "No Ids was provided", null);
@@ -130,10 +131,11 @@ public class ArtistLoader extends AbstractLoader {
 
         else{
             sortOrder = parseSortOrder(sortType);
+            selection = selection + " =?";
             selectionArgs = new String[]{ ids.get(0) };
         }
 
-        createLoadTask(result, MediaStore.Audio.Artists._ID, selectionArgs,
+        createLoadTask(result, selection, selectionArgs,
                 sortOrder, QUERY_TYPE_DEFAULT).execute();
     }
     /**

--- a/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/GenreLoader.java
+++ b/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/GenreLoader.java
@@ -19,7 +19,9 @@ import io.flutter.plugin.common.MethodChannel;
 public class GenreLoader extends AbstractLoader {
 
     private static final String[] GENRE_PROJECTION = {
-            "genre_name",
+//            "name",
+            MediaStore.Audio.GenresColumns.NAME,
+           // MediaStore.Audio.Genres._ID,
             //MediaStore.Audio.GenresColumns.NAME,
     };
 
@@ -50,7 +52,7 @@ public class GenreLoader extends AbstractLoader {
 
             default:
             case DEFAULT:
-                sortOrder = "genre_name ASC";
+                sortOrder = GENRE_PROJECTION[0] + " ASC";
                 //sortOrder = MediaStore.Audio.Genres.DEFAULT_SORT_ORDER;
                 break;
         }
@@ -111,7 +113,7 @@ public class GenreLoader extends AbstractLoader {
             Cursor genreCursor = null;
 
             try {
-                genreCursor = m_resolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
+                genreCursor = m_resolver.query(MediaStore.Audio.Genres.EXTERNAL_CONTENT_URI,
                         new String[]{"Distinct " + GENRE_PROJECTION[0]}, selection,
                         selectionArgs, sortOrder);
 

--- a/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/GenreLoader.java
+++ b/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/GenreLoader.java
@@ -108,16 +108,15 @@ public class GenreLoader extends AbstractLoader {
                                                      String sortOrder) {
 
             List<Map<String, Object>> dataList= new ArrayList<>();
-            Cursor genreCursor = m_resolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
-                    new String[] {"Distinct " + GENRE_PROJECTION[0] }, selection,
-                    selectionArgs, sortOrder);
+            Cursor genreCursor = null;
 
-//            Cursor genreCursor = m_resolver.query(MediaStore.Audio.Genres.EXTERNAL_CONTENT_URI,
-//                    GENRE_PROJECTION, selection, selectionArgs, sortOrder);
+            try {
+                genreCursor = m_resolver.query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI,
+                        new String[]{"Distinct " + GENRE_PROJECTION[0]}, selection,
+                        selectionArgs, sortOrder);
 
-            if (genreCursor != null){
-                while ( genreCursor.moveToNext() ){
-                    try {
+                if (genreCursor != null) {
+                    while (genreCursor.moveToNext()) {
                         Map<String, Object> data = new HashMap<>();
                         for (String column : genreCursor.getColumnNames()) {
                             String genreName = genreCursor.getString(
@@ -126,13 +125,13 @@ public class GenreLoader extends AbstractLoader {
                         }
                         dataList.add(data);
                     }
-
-                    catch(Exception ex){
-                        Log.e(TAG_ERROR, "GenreLoader::loadData method exception");
-                        Log.e(TAG_ERROR, ex.getMessage() );
-                    }
+                    genreCursor.close();
                 }
-                genreCursor.close();
+            }
+
+            catch(RuntimeException ex){
+                Log.e(TAG_ERROR, "GenreLoader::loadData method exception");
+                //Log.e(TAG_ERROR, ex.getMessage() );
             }
 
             return dataList;

--- a/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/SongLoader.java
+++ b/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/SongLoader.java
@@ -257,14 +257,14 @@ public class SongLoader extends AbstractLoader {
     /**
      * This method queries songs from a specific artist.
      * @param result MethodChannel.Result object to send reply for dart.
-     * @param artistName Artist name that we want fetch songs.
+     * @param artistId Artist name that we want fetch songs.
      * @param sortType SongSortType object to define sort type for data queried.
      */
-    public void getSongsFromArtist(final MethodChannel.Result result, final String artistName,
+    public void getSongsFromArtist(final MethodChannel.Result result, final String artistId,
                                    final SongSortType sortType ){
 
-        createLoadTask(result, MediaStore.Audio.Media.ARTIST + " =?",
-                new String[] { artistName }, parseSortOrder(sortType), QUERY_TYPE_DEFAULT )
+        createLoadTask(result, MediaStore.Audio.Media.ARTIST_ID + " =?",
+                new String[] { artistId }, parseSortOrder(sortType), QUERY_TYPE_DEFAULT )
                 .execute();
     }
 

--- a/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/SongLoader.java
+++ b/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/SongLoader.java
@@ -294,6 +294,7 @@ public class SongLoader extends AbstractLoader {
 
         String[] selectionArgs;
         String sortOrder = null;
+        String selection = MediaStore.Audio.Media._ID;
 
         if (ids == null || ids.isEmpty()) {
             result.error("NO_SONG_IDS", "No Ids was provided", null);
@@ -309,10 +310,11 @@ public class SongLoader extends AbstractLoader {
 
         else{
             sortOrder = parseSortOrder(sortType);
+            selection = selection + " =?";
             selectionArgs = new String[]{ ids.get(0) };
         }
 
-        createLoadTask(result, MediaStore.Audio.Media._ID, selectionArgs,
+        createLoadTask(result, selection, selectionArgs,
                 sortOrder, QUERY_TYPE_DEFAULT).execute();
     }
 

--- a/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/SongLoader.java
+++ b/android/src/main/java/boaventura/com/devel/br/flutteraudioquery/loaders/SongLoader.java
@@ -1,8 +1,11 @@
 package boaventura.com.devel.br.flutteraudioquery.loaders;
 
 import android.content.ContentResolver;
+import android.content.ContentUris;
 import android.content.Context;
 import android.database.Cursor;
+import android.net.Uri;
+import android.os.Build;
 import android.provider.MediaStore;
 import android.util.Log;
 
@@ -18,7 +21,6 @@ import io.flutter.plugin.common.MethodChannel;
 public class SongLoader extends AbstractLoader {
 
     private static final String TAG = "MDBG";
-    //private static final String GENRE_NAME = "genre_name";
 
     private static final int QUERY_TYPE_GENRE_SONGS = 0x01;
     private static final int QUERY_TYPE_ALBUM_SONGS = 0x02;
@@ -461,11 +463,16 @@ public class SongLoader extends AbstractLoader {
 
                 while( songsCursor.moveToNext() ){
                     try {
-
                         Map<String, Object> songData = new HashMap<>();
-
                         for (String column : songsCursor.getColumnNames()){
                             switch (column ){
+                                case MediaStore.Audio.Media._ID:
+                                    String id = songsCursor.getString( songsCursor.getColumnIndex(column));
+                                    final Uri uri = ContentUris.appendId(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI.buildUpon(),
+                                            Long.parseLong( id )).build();
+                                    songData.put("uri" , uri.toString() );
+                                    break;
+                                    
                                 case MediaStore.Audio.Media.IS_MUSIC:
                                 case MediaStore.Audio.Media.IS_PODCAST:
                                 case MediaStore.Audio.Media.IS_RINGTONE:

--- a/lib/src/flutter_audio_query.dart
+++ b/lib/src/flutter_audio_query.dart
@@ -187,12 +187,12 @@ class FlutterAudioQuery {
 
   /// This method returns list with  all songs info from a specific artist.
   /// using his name.
-  /// [artist] Artist name must be non null
+  /// [artistId] Artist id must be non null
   Future<List<SongInfo>> getSongsFromArtist(
-      {@required final String artist,
+      {@required final String artistId,
       SongSortType sortType = SongSortType.DEFAULT}) async {
     List<dynamic> dataList = await channel.invokeMethod("getSongsFromArtist", {
-      'artist': artist,
+      'artist': artistId,
       SOURCE_KEY: SOURCE_SONGS,
       SORT_TYPE: sortType.index,
     });

--- a/lib/src/song_info.dart
+++ b/lib/src/song_info.dart
@@ -46,6 +46,8 @@ class SongInfo extends DataModel {
   /// Returns a String with a file path to audio data file
   String get filePath => _data['_data'];
 
+  String get uri => _data["uri"];
+
   /// Returns a String with the size, in bytes, of this audio file.
   String get fileSize => _data['_size'];
 

--- a/pubspec.yaml
+++ b/pubspec.yaml
@@ -1,7 +1,7 @@
 name: flutter_audio_query
 description: Flutter plugin to query data about artists, albums, songs, genres and playlists
   from device storage.
-version: 0.3.4+2
+version: 0.3.5+5
 #author: Marcos Antonio Boaventura <scavenger.gnu@gmail.com>
 homepage: https://github.com/sc4v3ng3r/flutter_audio_query
 

